vpr/src/place/timing/PlacerTimingCosts.h

```diff
@@ -0,0 +1,242 @@
+         *
+         * Useful for client code operating on the cost values (e.g. difference between costs).
+         */
+        operator double() const {
+            return connection_cost_;
+        }
+
+      private:
+        PlacerTimingCosts* timing_costs_;
+        double& connection_cost_;
+    };
+
+    /**
+     * @brief Proxy class representing the connection costs of a net.
+     *
+     * Supports indexing by pin index to retrieve the ConnectionProxy for that pin/connection.
+     */
+    class NetProxy {
+      public:
+        NetProxy(PlacerTimingCosts* timing_costs, double* net_sink_costs)
+            : timing_costs_(timing_costs)
+            , net_sink_costs_(net_sink_costs) {}
+
+        ///@brief Indexes into the specific net pin/connection.
+        ConnectionProxy operator[](size_t ipin) {
+            return ConnectionProxy(timing_costs_, net_sink_costs_[ipin]);
+        }
+
+        const ConnectionProxy operator[](size_t ipin) const {
+            return ConnectionProxy(timing_costs_, net_sink_costs_[ipin]);
+        }
+
+      private:
+        PlacerTimingCosts* timing_costs_;
+        double* net_sink_costs_;
+    };
+
+    ///@brief Indexes into the specific net.
+    NetProxy operator[](ClusterNetId net_id) {
+        VTR_ASSERT_SAFE(net_start_indices_[net_id] >= 0);
+
+        double* net_connection_costs = &connection_costs_[net_start_indices_[net_id]];
+        return NetProxy(this, net_connection_costs);
+    }
+
+    NetProxy operator[](ClusterNetId net_id) const {
+        VTR_ASSERT_SAFE(net_start_indices_[net_id] >= 0);
+
+        const double* net_connection_costs = &connection_costs_[net_start_indices_[net_id]];
+        return NetProxy(const_cast<PlacerTimingCosts*>(this), const_cast<double*>(net_connection_costs));
+    }
+
+    void clear() {
+        connection_costs_.clear();
+        net_start_indices_.clear();
+    }
+
+    void swap(PlacerTimingCosts& other) {
+        std::swap(connection_costs_, other.connection_costs_);
+        std::swap(net_start_indices_, other.net_start_indices_);
+        std::swap(num_levels_, other.num_levels_);
+    }
+
+    /**
+     * @brief Calculates the total cost of all connections efficiently
+     *        in the face of modified connection costs.
+     */
+    double total_cost() {
+        float cost = total_cost_recurr(0); //Root
+
+        VTR_ASSERT_DEBUG_MSG(cost == total_cost_from_scratch(0),
+                             "Expected incremental and from-scratch costs to be consistent");
+
+        return cost;
+    }
+
+  private:
+    ///@brief Recursively calculate and update the timing cost rooted at inode.
+    double total_cost_recurr(size_t inode);
+
+    double total_cost_from_scratch(size_t inode) const;
+
+    ///@brief Friend-ed so it can call invalidate().
+    friend ConnectionProxy;
+
+    void invalidate(const double* invalidated_cost);
+
+    size_t left_child(size_t i) const {
```

Comment:
I would make this and the following methods static. This just makes it more clear that these are not dependent on the instance of the class.